ucca/convert.py

```diff
@@ -879,51 +907,54 @@ def from_json(lines, *args, skip_category_mapping=False, by_external_id=False, *
+        children_tokens = [] if unit["type"] == "IMPLICIT" else unit["children_tokens"]
+        try:
+            terminal = token_id_to_terminal[children_tokens[0]["id"]] if len(children_tokens) == 1 else None
+        except (IndexError, KeyError):
+            terminal = None
+        if remote:
             try:
-                terminal = token_id_to_terminal[children_tokens[0]["id"]] if len(children_tokens) == 1 else None
-            except (IndexError, KeyError):
-                terminal = None
-            if remote:
-                try:
-                    node = tree_id_to_node[cloned_from_tree_id]
-                except KeyError as e:
-                    raise ValueError("Remote copy %s refers to nonexistent unit: %s" %
-                                     (tree_id, cloned_from_tree_id)) from e
-                l1.add_remote(parent_node, tag, node)
-            elif not skip_category_mapping and terminal and layer0.is_punct(terminal):
-                tree_id_to_node[tree_id] = l1.add_punct(None, terminal)
-            elif tree_id not in tree_id_to_node:
-                node = tree_id_to_node[tree_id] = l1.add_fnode(parent_node, tag, implicit=(unit["type"] == "IMPLICIT"))
-                node.extra['tree_id'] = tree_id
-                comment = unit.get("comment")
-                node.extra["all_tags"] = ';'.join(tags)
-                if comment:
-                    node.extra['remarks'] = comment
-                for token in children_tokens:
-                    token_id_to_preterminal[token["id"]] = node
-            # currently only supports one non-ignored category, TODO: fix it
+                node = tree_id_to_node[cloned_from_tree_id]
+            except KeyError as e:
+                raise ValueError("Remote copy %s refers to nonexistent unit: %s" %
+                                 (tree_id, cloned_from_tree_id)) from e
+            l1.add_remote(parent_node, unit_categories, node)
+        elif not skip_category_mapping and terminal and layer0.is_punct(terminal):
+            tree_id_to_node[tree_id] = l1.add_punct(None, terminal, base_layer, base_slot)
+        elif tree_id not in tree_id_to_node:
+            node = tree_id_to_node[tree_id] = l1.add_fnode(parent_node, unit_categories, implicit=(unit["type"] == "IMPLICIT"))
+            node.extra['tree_id'] = tree_id
+            comment = unit.get("comment")
+            if comment:
+                node.extra['remarks'] = comment
+            for token in children_tokens:
+                token_id_to_preterminal[token["id"]] = node
 
     # Attach terminals to non-terminals
     for token_id, node in token_id_to_preterminal.items():
         terminal = token_id_to_terminal[token_id]
         if skip_category_mapping or not layer0.is_punct(terminal):
-            node.add(EdgeTags.Terminal, terminal)
+            node.add([(EdgeTags.Terminal, base_slot, base_layer, "")], terminal)
```

Comment:
`""` for parent can be omitted since it's optional (same for other places in the code)